### PR TITLE
Set `maxUnavailable = 0` for all deployments

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR is a replacement for https://github.com/grafana/mimir/pull/5017, based on @pracucci's suggestion from https://github.com/grafana/mimir/pull/5017#pullrequestreview-1593029956.

This PR sets `maxUnavailable` to 0 and `maxSurge` to 15% by default for all deployments. This ensures we don't create a situation during a rollout where we have no healthy pods for a deployment. With the previous configuration, if Kubernetes stops the only healthy pod, then tries to start a new pod with the new configuration and that pod cannot start, we'll have no healthy pods for the deployment.

This PR also extends our OPA policies to ensure that deployments have `maxUnavailable` set to 0 in the future.

Applying this change to most deployments seemed straightforward, but there were a few that I'm open to feedback on:

* `rollout-operator`: doesn't support running multiple instances at once, so I've set `maxSurge` to 0 and `maxUnavailable` to 1, and excluded it from the new OPA policy

* `consul` and `overrides-exporter`: I've set `maxUnavailable` to 0 and `maxSurge` to 15%, but I'm unsure if running multiple Consul or overrides-exporter instances simultaneously is a good idea (previous configuration was using Kubernetes' defaults, which is 25% for both `maxUnavailable` and `maxSurge`)

* `query-scheduler`: I've set `maxSurge` to 1, rather than 15%, to ensure we don't end up with many schedulers running in parallel (previous configuration was `maxUnavailable` set to 1 and `maxSurge` set to 0) 

* `continuous-test`: I don't think it makes sense to run multiple instances of this in parallel, and it being unavailable only affects operators of Mimir, not end users, so I've set `maxSurge` to 0 and `maxUnavailable` to 1, and excluded it from the new OPA policy

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated
- [n/a] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: operations/helm/charts/mimir-distributed/CHANGELOG.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ Entries should include a reference to the Pull Request that introduced the chang
 * [ENHANCEMENT] Sets the `appProtocol` value to `tcp` for the `gossip-ring-svc` service template. This allows memberlist to work with istio protocol selection. #5673
 * [ENHANCEMENT] Update the `rollout-operator` subchart to `0.8.0`. #5718
 * [ENHANCEMENT] Make store_gateway podManagementPolicy configurable. #5757
+* [ENHANCEMENT] Set `maxUnavailable` to 0 for most deployments, to ensure they don't become completely unavailable during a rollout. #5924
```
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/CHANGELOG.md:36]: nitpick but can you list all the deployments this affects? or all the deployments it doesn't affect?

File: operations/policies/deployment_rollingupdate.rego [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	maxUnavailable := obj.spec.strategy.rollingUpdate.maxUnavailable
+	msg = sprintf("%s has spec.strategy.rollingUpdate.maxUnavailable set to %v, but 0 is required", [object_display_name[i], maxUnavailable])
+
+	obj.kind == "Deployment"
+	not is_ignored_deployment(obj)
+	maxUnavailable > 0
+}
+
+is_ignored_deployment(obj) {
+	obj.metadata.name in ignored_deployments
```
Comment (dimitarvdimitrov) [inline operations/policies/deployment_rollingupdate.rego:48]: Does this work as expected? Aren't lot of the actual names are prefixed with the release name of the chart? e.g. `test-mimir-continuous-test`? I think the problem may be coming from the fact that we don't render continuous test in any of the offline tests.

File: operations/policies/deployment_rollingupdate.rego [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package main
+
+import future.keywords.in
+
+ignored_deployments = [
+	"continuous-test", # We only want to run at most one instance at a time.
+	"rollout-operator", # We only want to run at most one instance at a time.
```
Comment (dimitarvdimitrov) [inline operations/policies/deployment_rollingupdate.rego:7]: We don't run the rego policies on subcharts, so this won't have any effect. I think we should remove this from the exclusions.

**Review by dimitarvdimitrov — COMMENTED** [review]
mostly LGTM, thanks for doing the rego policies changes!

File: operations/policies/deployment_rollingupdate.rego (hunk 3 above)
Comment (charleskorn) [reply operations/policies/deployment_rollingupdate.rego:7]: This policy is also applied to the rendered Jsonnet manifests, which does include `rollout-operator`.

File: operations/policies/deployment_rollingupdate.rego (hunk 2 above)
Comment (charleskorn) [reply operations/policies/deployment_rollingupdate.rego:48]: The Rego policies are also applied to the rendered Jsonnet manifests, where this does apply.

File: operations/helm/charts/mimir-distributed/CHANGELOG.md (hunk 1 above)
Comment (charleskorn) [reply operations/helm/charts/mimir-distributed/CHANGELOG.md]: Good idea, added in 1e99303.

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM; for some reason gmail thought that emails for this PR are spam, so I only now saw your response

**Comment by charleskorn** [comment]
> LGTM; for some reason gmail thought that emails for this PR are spam, so I only now saw your response

It thought your initial review was spam as well 🤷 Maybe Gmail doesn't want us to have reliable deployments...

**Pull request merged**